### ISSUE TITLE
Negative Loss while training the model

### ISSUE THREAD

**Issue by mansiag** [cpmpercussion/keras-mdn-layer#15]
Hi! I was training a mdn model with 4 outputs but after 2-3 epochs, the loss became negative. Is it possible that the loss can be negative ?

**Comment by cpmpercussion** [comment]
Yes the loss can be negative and it’s ok! Weird huh? Lower is still better.

It’s because the loss is calculated using the probability density function of a continuous distribution which works differently to the PDF for a discrete distribution. These functions can end up higher than 1 at some points (leading to less that 0 loss), as long as the area under the curve is equal to one.

You can try this if you like with the PDF for a Gaussian distribution with a very small standard deviation. As the sd approaches zero, the PDF goes to infinity at the mean.

It’s bad if the PDF in the loss function goes to infinity, so we add a small number to the standard deviations in our calculations to make sure that doesn’t happen.

**Comment by mansiag** [comment]
So, getting a negative loss is good , that is should I continue training on
getting negative loss..moving it towards more negative ? Or should I stop
the training before 0 loss.

**Comment by cpmpercussion** [comment]
Keep training to see how low it will go! Lower is better and 0 has no particular meaning with this loss function.
